File: api/models/ClassPurchased.js
var mongoose = require('mongoose');
var Schema = mongoose.Schema;

var ClassPurchasedSchema = new Schema({

    //Class ID
    ClassID: {
        type: mongoose.Schema.Types.ObjectId,
        ref: 'Class',
        required: true,
    },

    //Purchased User ID
    UserID: {
        type: mongoose.Schema.Types.ObjectId,
        ref: 'User',
        required: true,
    },

    //Class Schedules
    ClassTimes: [{
        //Start Date
        StartDate: {
            type: Date,
        },

        //End Date
        EndDate: {
            type: Date,
        },


        //Recurrence Type
        // Recurrence: {
        //     type: [{

        //     type: String,
        //         enum: ['None', 'Daily', 'Weekly', 'BiWeekly', 'Monthly', 'Yearly']
        //         }],

        //     default: ['None'],
        // },
    }],

    //Class Schedules that the user decides to update
    UpdatedClassTimes: [{
        //Start Date
        StartDate: {
            type: Date,
        },

        //End Date
        EndDate: {
            type: Date,
        },

        ScheduleReference: {
            type: mongoose.Schema.Types.ObjectId,
            required: true,
        }
    }],

    //Class Schedules that the user decides to cancel
    CancelledClassTimes: [{
        //Start Date
        StartDate: {
            type: Date,
        },

        //End Date
        EndDate: {
            type: Date,
        },

        ScheduleReference: {
            type: mongoose.Schema.Types.ObjectId,
            required: true,
        }
    }],

    //Date Booked
    DateBooked: {
        type: Date,
        required: true,
    },

    //Price Paid
    PricePaid: {
        type: Number,
        required: true,
    },

    //Did trainee miss class
    IsMissed: {
        type: Boolean
    },

    //Did trainee cancell class
    IsCancelled: {
        type: Boolean
    },

    //Trainee's cancellation reason
    CancellationReason: {
        type: String
    }
})

module.exports = mongoose.model('Class Purchased', ClassPurchasedSchema)

File: lib/feModels/ClassPurchased.dart
// ignore: constant_identifier_names
import 'package:balance/feModels/ScheduleModel.dart';

enum RecurrenceType { None, Daily, Weekly, BiWeekly, Monthly, Yearly }

// ignore: constant_identifier_names
enum ClassType { Solo, Group, Virtual }

class ClassSchedule {
  String classID;
  String userID;

  late List<Schedule> classTimes;
  List<UpdatedSchedule>? updatedClassTimes;
  List<CancelledSchedule>? cancelledClassTimes;

  DateTime dateBooked;
  double pricePaid;
  bool? isMissed;
  bool? isCancelled;
  String? cancellationReason;

  ClassSchedule(
      {required this.classID,
      required this.userID,
      required this.dateBooked,
      required this.pricePaid});
}
